### PR TITLE
Adding length check for AVI objects

### PR DESCRIPTION
**Pull request by akshayhavile**
This PR deals with length check for AVI Objects in SNI and EVH deployment. Layer 2 and Layer 3 checks.
Scenario:
1. New deployment: This code will not allow to create a new VS if name of any of the object associated with VS exceed max length criteria. Once ingress/route corrected, new VS will be created.
2. Existing Deployment: If existing deployment upgrade to this build already contain object whose name exceeds max length criteria, AKO will just print warning message in logs (layer 2 checks). Customer has to correct that VS manually. 

This PR does not take care of following things:
1. Deleting objects from AVI side which are in error states.

@apalsule  @sudswasavi  @monotosh-avi  @chauhanshubham  @hemantavi : Few observation during testings:
Correct State: Ingress with all fields proper (less than equal to length restriction)
Incorrect State: Ingress with few/all fields improper (greater than length restriction)
1. Insecure ingress SNI deployment: If customer update ingress from correct to incorrect state (not updating FQDN here), previous pools will be deleted. But no improper update at AVI Controller side. Delete improper ingress from K8 side results in deleting all info associated with ingress getting removed from VS (as pools are already deleted, no new pool created, so only dns info will be deleted from AVI side).
2. Secure SNI, Insecure and Secure EVH: If customer update ingress from correct state to incorrect state (not updating FQDN here),  no update happens at AVI Controller side. Delete of improper ingress does not delete previously created objects from controller side. Reason: PG of VS not empty so model deletion doesn't happen. AKO needs to be rebooted for that.

### CODE REVIEW

**Comment by akshayhavile** [comment]
build ako

**Comment by akshayhavile** [comment]
build ako

**Comment by akshayhavile** [comment]
build ako

**Comment by akshayhavile** [comment]
build ako

File: internal/rest/avi_obj_httpps.go [hunk 1]
```diff
@@ -49,116 +55,133 @@ func (rest *RestOperations) AviHttpPSBuild(hps_meta *nodes.AviHttpPolicySetNode,
...
-		if len(hppmap.Hosts) > 0 {
-			match_crit := "HDR_EQUALS"
-			host_hdr_match := avimodels.HostHdrMatch{MatchCriteria: &match_crit,
-				Value: hppmap.Hosts}
-			match_target.HostHdr = &host_hdr_match
-			port_match_crit := "IS_IN"
-			match_target.VsPort = &avimodels.PortMatch{MatchCriteria: &port_match_crit, Ports: []int64{int64(hppmap.VsPort)}}
+		if len(name) > lib.AVI_OBJ_NAME_MAX_LENGTH {
+			utils.AviLog.Warnf("key: %s, msg: HTTPS: length of HTTP Redirect Rule: %s exceeds max length limit for AVI Objects. Not adding rule to HTTPS object",
+				key, name)
```
Comment (monotosh-avi) [inline internal/rest/avi_obj_httpps.go:127]: nit: we can add continue here and avoid next else statement

**Review by monotosh-avi — DISMISSED** [review]

**Comment by akshayhavile** [comment]
build ako

**Comment by akshayhavile** [comment]
build ako

**Comment by chauhanshubham** [comment]
build ako

**Comment by chauhanshubham** [comment]
build ako

File: internal/lib/lib.go [hunk 2]
```diff
@@ -176,34 +185,40 @@ func GetModelName(namespace, objectName string) string {
 
 // All L4 object names.
 func GetL4VSName(svcName, namespace string) string {
-	return Encode(NamePrefix + namespace + "-" + svcName)
+	return Encode(NamePrefix+namespace+"-"+svcName, "L4 Virtual Service ")
```
Comment (chauhanshubham) [inline internal/lib/lib.go:188]: why the trailing space in the end of the objType strings? also should we keep it single words or better constants?

File: internal/rest/avi_datascript.go [hunk 3]
```diff
@@ -29,6 +29,12 @@ import (
 )
 
 func (rest *RestOperations) AviDSBuild(ds_meta *nodes.AviHTTPDataScriptNode, cache_obj *avicache.AviDSCache, key string) *utils.RestOp {
+
+	if len(ds_meta.Name) > lib.AVI_OBJ_NAME_MAX_LENGTH {
```
Comment (chauhanshubham) [inline internal/rest/avi_datascript.go:33]: `PrintObjectNameLengthWarn` can be called here.

File: internal/rest/avi_evh_obj.go [hunk 4]
```diff
@@ -215,12 +223,20 @@ func (rest *RestOperations) EvhNodeCU(sni_node *nodes.AviEvhVsNode, vs_cache_obj
...
+		if restOp != nil {
+			rest_ops = append(rest_ops, restOp...)
+		}
 	}
 	return cache_sni_nodes, rest_ops
 }
 
 func (rest *RestOperations) AviVsBuildForEvh(vs_meta *nodes.AviEvhVsNode, rest_method utils.RestMethod, cache_obj *avicache.AviVsCache, key string) []*utils.RestOp {
+
+	if len(vs_meta.Name) > lib.AVI_OBJ_NAME_MAX_LENGTH {
```
Comment (chauhanshubham) [inline internal/rest/avi_evh_obj.go:235]: same here, we can use the `PrintObjectNameLengthWarn` function here, and in other build functions too.

File: tests/evhtests/l7_evh_crd_test.go [hunk 5]
```diff
@@ -381,7 +381,7 @@ func TestInsecureHostAndHostruleForEvh(t *testing.T) {
 		nodes := aviModel.(*avinodes.AviObjectGraph).GetAviEvhVS()
 		return len(nodes[0].EvhNodes)
 	}, 10*time.Second).Should(gomega.Equal(1))
-	sniVSKey := cache.NamespaceName{Namespace: "admin", Name: lib.Encode("cluster--default-foo.com")}
+	sniVSKey := cache.NamespaceName{Namespace: "admin", Name: lib.Encode("cluster--default-foo.com", "SNI VS ")}
```
Comment (chauhanshubham) [inline tests/evhtests/l7_evh_crd_test.go:384]: I think it might be good to use constants everywhere.

**Comment by akshayhavile** [comment]
build ako

**Review by chauhanshubham — APPROVED** [review]

**Review by monotosh-avi — APPROVED** [review]

**Pull request merged**